### ISSUE TITLE
How do I include node_modules css files?

### ISSUE THREAD

**Issue by simkessy** [webpack/webpack#1789]
I'm having a hard time figuring this out. I have all these node_modules which include the css files I need but I can only get the javascript from the require. 

For example require('bootstrap') will only return the js file and not the css. 

I've tried 

```
require("!style!css!bootstrap/dist/css");
``` 
but I just get this error:

>ERROR in ./app/js/loader.js
Module not found: Error: Cannot resolve module 'bootstrap.css' in C:\Users\Administrator\Dropbox\projects\spDash\app\js
 @ ./app/js/loader.js 10:0-24

>ERROR in ./app/js/loader.js
Module not found: Error: Cannot resolve module 'bootstrap/dist/css' in C:\Users\Administrator\Dropbox\projects\spDash\app\js
 @ ./app/js/loader.js 19:0-40

My config looks like this:

```
var webpack = require('webpack')
module.exports = {
  entry: {
    'spdash': './app/js/loader.js'
  },
  output: {
    filename: 'bundle.js',
  },
  devtool: 'source-map',
  module: {
    loaders: [
      // the url-loader uses DataUrls.
      // the file-loader emits files.
      { test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: "url-loader?limit=10000&minetype=application/font-woff" },
      { test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/, loader: "file-loader" },
      {
        test: /\.css$/, // Only .css files
        loader: 'style!css' // Run both loaders
      },
      { test: /\.less$/, loader: 'style-loader!css-loader!less-loader' }, // use ! to chain loaders
      { test: /\.css$/, loader: 'style-loader!css-loader' },
      {test: /\.(png|jpg)$/, loader: 'url-loader?limit=8192'} // inline base64 URLs for <=8k images, direct URLs for the rest
    ]
  },
  plugins: [
    new webpack.optimize.UglifyJsPlugin({
      minimize: true,
      compress: {
        warnings: false
      }
    }),
    new webpack.IgnorePlugin(/^\.\/locale$/, [/moment$/])
  ]
};
```

**Comment by silvenon** [comment]
Try `require("!style!css!bootstrap/dist/css/bootstrap.css")`. Also, you don't need `!style!css!` since you already have that specified in the loaders configuration.

**Comment by simkessy** [comment]
Hey this can be closed, I just needed more experience with webpack I've since figured it out.

**Comment by ArdentKid** [comment]
I still need help... how can I include the css from a node_modules package??

**Comment by silvenon** [comment]
@ArdentKid it's same as including a local CSS file. You include npm modules with `require('some-module')`, so if that module has a CSS file somewhere, you would do `require('some-module/path/to/style.css')`. Make sure you have a CSS loader defined.

**Comment by lfender6445** [comment]
another option here is to provide a resolve alias

```js
    resolve: {
      modules: ['node_modules', 'src'],
      alias: {
        'react-input-range-css': path.join(__dirname, '../node_modules/react-input-range/dist/react-input-range.css')
      }
```

aliasing allows you to rename modules https://webpack.github.io/docs/resolving.html#aliasing

**Comment by MatthewKosloski** [comment]
I'm trying to import `normalize-css` and `bourbon` into Webpack; however, I am experiencing difficulty.  

A brief summary of my file structure:

	root-folder
		node_modules
		...
			bourbon
				app
					assets
						stylesheets
							_bourbon.scss
			normalize-css
				normalize.css
		...
		src
			components
				Foo
					index.js
					style.scss
			scss
				lib
					bourbon.scss
		webpack.config.js


_./src/components/Foo/style.scss:_

     @import '../../../node_modules/bourbon/app/assets/stylesheets/bourbon'; // WORKS FINE
     @import 'bourbon'; // DOES NOT WORK

_./src/scss/lib/bourbon.scss:_

    @import '../../../node_modules/bourbon/app/assets/stylesheets/bourbon';

_./webpack.config.js:_

    const webpack = require('webpack');
    const path = require('path');
    const ExtractTextPlugin = require('extract-text-webpack-plugin');

    // require('normalize-css/normalize.css');

    module.exports = {
        devtool: 'source-map',
        entry: {
            bundle: './src/app'
        },
        output: {
            path: path.join(__dirname, 'public'),
            filename: './js/[name].js',
            publicPath: '/static'
        },
        module: {
            loaders: [
                {
                    test: /\.js$/,
                    exclude:  /node_modules/,
                    loader: 'babel'
                },
                {
                    test: /\.css$/,
                    include: /node_modules/,
                    loader:  'style!css'
                },
                {
                    test: /\.scss$/,
                    exclude:  /node_modules/,
                    loader:  ExtractTextPlugin.extract('style', 'css?modules&localIdentName=[name]__[local]___[hash:base64:5]!sass')
                }
            ]
        },
        plugins: [
            new webpack.optimize.OccurrenceOrderPlugin(),
            new ExtractTextPlugin('./css/[name].css', {allChunks: true})
        ],
        resolve: {
            extensions: ['', '.js', '.scss'],
            alias: {
                bourbon: path.join(__dirname, 'node_modules/bourbon/app/assets/stylesheets/bourbon')
            }
        }
    };

When I try to use `resolve.alias.bourbon` within _./src/components/Foo/style.scss_, the following error is logged:

        ERROR in ./~/css-loader?modules&localIdentName=[name]__[local]___[hash:base64:5]!./~/sass-loader!./src/components/Foo/style.scss
    Module build failed: 
    @import 'bourbon';
    ^
          File to import not found or unreadable: bourbon

` @import '../../../node_modules/bourbon/app/assets/stylesheets/bourbon';` works fine, but that is tedious and something tells me that it's bad practice.

When I try `require('normalize-css/normalize.css')`, I get the following error:

	summary {
	        ^
	SyntaxError: Unexpected token {	

It's as if the css loader didn't recognize the file...

**Comment by silvenon** [comment]
`@import 'bourbon'` won't work because a) instead of the main SCSS, `bourbon` has [`index.js` defined as `main`](https://github.com/thoughtbot/bourbon/blob/master/package.json#L23) (not sure what that `index.js` is supposed to be), and b) because modules should be imported with the [`~` prefix](https://github.com/webpack/css-loader#usage), try `@import '~bourbon/core/_bourbon.scss'`.

Concerning `normalize-css`, try installing [normalize.css](https://www.npmjs.com/package/normalize.css) instead and doing `require('normalize.css')`.

Let me know if that works.

**Comment by MatthewKosloski** [comment]
Thank you @silvenon for replying so quickly.  I tried what you suggested, but no results :/

I installed normalize.css and added `require('normalize.css')` to my `webpack.config.js`, but I still get the "Unexpected token" error.

When I tried `@import '~bourbon/core/_bourbon.scss'` in _./src/components/Foo/style.scss:_, I still get the " File to import not found or unreadable: bourbon" error.  It kinda sucks that this isn't documented anywhere.

**Comment by MatthewKosloski** [comment]
Is there a way to import a file only once _or_ change the order?  For example, Normalize.css.  I want it at the very beginning of the cascade.  I figured out how to import it once by importing normalize at the top level component (app.js), but Webpack processes it last and it ends up in the bottom, which is not what I want.

_Consider the following file structure:_

	components
		Foo
			foo.scss
		Bar
			bar.scss
	app.js
	app.scss

_App.js is an entry point like so:_

	import React, { Component } from 'react';
	import { render } from 'react-dom';
	import Foo from './components/Foo';
	import Bar from './components/Bar';
	import s from './app.scss';

	const app = (
		<div>
			<Foo />
			<Bar />
		</div>
	);

	render(app, document.getElementById('app'));

_Here is app.scss:_

	@import '~normalize';

_Depiction of the outputted bundle, or the "cascade."  (I need Normalize on top):_

	Foo's rules (foo.scss)
	Bar's rules (bar.scss)
	Normalize.css (app.scss)



The reason why I need to "import once" is that the output includes normalize everywhere where I import it.  So, if I were to include it in both Foo and Bar, it would show up twice.  This goes for ANY css rule that I don't set globally like this example:

	:global(.clearfix) {
		&::after {
			content: '';
			display: table;
			clear: both;
		}
	}

And I obviously have no intent on manipulating a node module to add the `:global()` function on every Normalize.css rule.

**Comment by silvenon** [comment]
That's weird, webpack should compile only one instance of every module, even if it's been imported multiple times into different files.

I see the problem with Normalize.css ending up at the bottom, that problem never occurred to me. You can't count on the order of `import` statements anyway. What you can do is import Normalize.css from your webpack config:

```js
{
  // ...
  entry: ['normalize', 'src/App.js']
  // ...
}
```

This way you can ensure the order. This is also a good place to import babel-polyfill if you're using it.

**Comment by MatthewKosloski** [comment]
@silvenon Thank you for your quick responses and willingness to help me—really appreciate it!  Unfortunately, Webpack is not easy to get up and running the way you want.

I figured out how to solve the duplicated code problem.  All I did was install a Webpack plugin called [optimize-css-assets-webpack-plugin](https://www.npmjs.com/package/optimize-css-assets-webpack-plugin).  This minifies CSS files and removes duplicate code.  Now, I can import anywhere I want without the worry; however, this doesn't solve the order of the CSS. :(

I tried adding Normalize as another entry in my config, but it didn't work.  I didn't get any errors, and the Normalize code wasn't showing up in `bundle.css`.

Here is my new latest config file:

    const webpack = require('webpack');
    const path = require('path');
    const ExtractTextPlugin = require('extract-text-webpack-plugin');
    const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin');

    const debug = process.env.NODE_ENV !== 'production';

    module.exports = {
        devtool: debug ? 'source-map' : null,
        entry: ['./node_modules/normalize.css', './src/app.js'],
        output: {
            path: path.join(__dirname, 'public'),
            filename: debug ? './js/bundle.js' : './js/bundle.min.js',
            publicPath: '/static'
        },
        module: {
            loaders: [
                {
                    test: /\.js$/,
                    exclude:  /node_modules/,
                    loader: 'babel'
                },
                {
                    test: /\.css$/,
                    loader:  'style!css'
                },
                {
                    test: /\.scss$/,
                    loader:  ExtractTextPlugin.extract('style', `css?modules&localIdentName=${debug ? 'bundle-scss__[local]-class___[hash:base64:3]' : 'mk-[hash:base64:5]'}!sass`)
                }
            ]
        },
        plugins: debug ? 
        [
            new webpack.optimize.OccurrenceOrderPlugin(),
            new ExtractTextPlugin('./css/bundle.css', {allChunks: true})
        ] : [
            new webpack.optimize.DedupePlugin(),
            new webpack.optimize.OccurenceOrderPlugin(),
            new webpack.optimize.UglifyJsPlugin({compress: {warnings: false}}),
            new ExtractTextPlugin('./css/bundle.min.css', {allChunks: true}),
            new OptimizeCssAssetsPlugin({
                assetNameRegExp: /\.css$/g,
                cssProcessor: require('cssnano'),
                cssProcessorOptions: { discardComments: {removeAll: true } },
                canPrint: true
            }),
            new webpack.DefinePlugin({
                'process.env':{
                    'NODE_ENV': JSON.stringify('production')
                }
            })
        ],
        resolve: {
            extensions: ['', '.js', '.css', '.scss'],
            alias: {
                main: path.join(__dirname, '/src/scss/main.scss'),
                bourbon: path.join(__dirname, '/node_modules/bourbon/app/assets/stylesheets/_bourbon.scss'),
                normalize: path.join(__dirname, '/node_modules/normalize.css')
            }
        }
    };

**Comment by silvenon** [comment]
> I didn't get any errors, and the Normalize code wasn't showing up in bundle.css.

It won't end up in the `bundle.css` because this is the loader you're using for CSS:

```js
{
  test: /\.css$/,
  loader:  'style!css'
},
```

Normalize should end up inside `<style>` tags on your page, please check if that happened. Because you're using CSS Modules, you'll want better control over which stylesheets are processed as CSS Modules and which aren't. I had this problem as well and this is what I've come up with, I modified it according to your config:

```js
// this part is completely up to you,
// I like to set up my environment like this
const IS_DEV = !process.env.NODE_ENV;
const IS_PROD = process.env.NODE_ENV === 'production';

// let's extract this logic here
const LOCAL_IDENT_NAME = IS_DEV ?
  'bundle-scss__[local]-class___[hash:base64:3]' :
  'mk-[hash:base64:5]';

// ...

{
  test: /\.s?css$/,
  loader: IS_DEV ?
    `style!css?modules&localIdentName=${LOCAL_IDENT_NAME}!sass` :
    ExtractTextPlugin.extract('style', 'css?modules&importLoaders=1!postcss!sass'),
  include: [
    // only stylesheets inside src/styles and the react-toolbox module
    // will be processed as CSS Modules
    resolve('./src/styles'),
    /react-toolbox/,
  ],
  exclude: [
    // EXCEPT for stylesheets in src/styles/global
    resolve('./src/styles/global'),
  ],
},
{
  test: /\.s?css$/,
  loader: IS_DEV ?
    'style!css!sass' :
    ExtractTextPlugin.extract('style', 'css!sass'),
  include: [
    // stylesheets in node_modules and src/styles/global
    // will be treated as global, i.e. not CSS Modules
    resolve('./node_modules'),
    resolve('./src/styles/global'),
  ],
  exclude: [
    // EXCEPT for react-toolbox, which should be treated as CSS Modules
    // as specified by the previous loader
    /react-toolbox/,
  ],
},
// ...
```

React Toolbox is just an example here, in case you ever have to install a module and treat it as CSS Modules.

As you said, webpack is not that easy. 😕

**Comment by MatthewKosloski** [comment]
Thank you for all your help, @silvenon! I adopted the "two loader approach" to match the files with the desired loader.  You were right about Normalize being rendered inline, too. Didn't catch that!

# **TL;DR**

Thank you!  If you'd like to see what I did, check out my [repository](https://github.com/MatthewKosloski/webpack).

# **Long explanation**

_Here is a brief summary of my new project setup:_

	root
		src
			components
				Foo
					foo.scss (`@import '~utils'`)
					index.js
			scss
				globals (Global styles that go before CSS module styles)
				utils (Bourbon, Sass mixins, variables, etc. that won't be outputted)
			app.js (entry point)

Here is the Webpack config file:

    ...

    const IS_DEV = process.env.NODE_ENV !== 'production';

    const LOCAL_IDENT_NAME = IS_DEV ? 
        '[name]-scss__[local]-class___[hash:base64:2]' 
    : '[hash:base64:2]';

    const globals = [
        resolve('./node_modules/normalize.css'), // any NPM module
        resolve('./src/scss/globals'), // global CSS classes
    ];

    const cssModules = [
        resolve('./src/components'), // CSS modules
    ];

    ...

    entry: {
        bundle: [
            './node_modules/normalize.css',
            './src/scss/globals',
            './src/app'
        ]
    },
    output: {
        path: path.join(__dirname, 'public'),
        filename: IS_DEV ? 'js/[name].js' : 'js/[name].min.js',
        publicPath: '/'
    },
    module: {
        loaders: [
            // loader for CSS modules
            {
                test: /\.s?css$/,
                loader:  ExtractTextPlugin.extract('style', `css?modules&localIdentName=${LOCAL_IDENT_NAME}!sass`),
                include: cssModules,
                exclude: globals
            },
            // loader for global styles
            {
                test: /\.s?css$/,
                loader:  ExtractTextPlugin.extract('style', 'css!sass'),
                include: globals,
                exclude: cssModules
            }
        ]
    },
    resolve: {
        extensions: ['', '.js', '.css', '.scss'],
        alias: {
            bourbon: path.join(__dirname, '/node_modules/bourbon/app/assets/stylesheets/_bourbon.scss'),
            utils: path.join(__dirname, '/src/scss/utils')
        }
    }

**Comment by brettstack** [comment]
For anyone trying to import node_modules packages from SCSS, prefix with `~`:

```
@import '~semantic-ui-css/semantic.css';
```

**Comment by sohaibjaved44** [comment]
I'm using aspnet core angular2 visual studio template and I have added primeNg module for datatable. Its working fine but for styling I need to add some references , I'm new to web pack so not sure where and how I can add them. 
Can anyone please give me hint?
Also if you can share some url to go through this process will be highly appreciated.
Thanks

**Comment by silvenon** [comment]
@sohaibjaved44 sounds like a question for Stack Overflow. I suggest being more specific when opening a question there, e.g. what your template looks like.

GitHub Issues is more for bringing bugs to attention, sending PRs, proposing new features and similar.

**Comment by danny-andrews-snap** [comment]
Is this behavior (requiring a tilde to import css from node modules) documented anywhere? I opened a [PR](https://github.com/webpack-contrib/css-loader/pull/777) over here which might help, but I don't think it's sufficient.

**Comment by silvenon** [comment]
It's documented [here](https://github.com/webpack-contrib/css-loader#url), but it's not very obvious.

**Comment by mattcarlotta** [comment]
For any one else who might have trouble with `webpack 4` while using `css-loader` with `css-modules` and are only using `scss` files, I was able to get this config to work:

**webpack.dev.js**
```
const resolve = require('path').resolve

const globalCSS = resolve(__dirname, '../', 'src/styles/globals'),

module.exports = {
  ...
  module: {
    rules: [
      {
        test: /\.s?css$/, // if stylesheet is CSS or SCSS
        exclude: [globalCSS], // exclude global styles (styles with node_module imports)
        use: [
          'style-loader',
          {
            loader: 'css-loader',
            options: {
              sourceMap: true,  // allow source mapping
              modules: true, // allow module imports (ex: import styles from 'example.scss')
              localIdentName: '[local]___[hash:base64:5]', // allow class hash renaming
            },
          },
          'sass-loader',
        ],
      },
      {
        test: /\.s?css$/, // if stylesheet is CSS or SCSS
        include: [globalCSS], // only include globally imported styles (ex: @import '~material-icons/iconfont/material-icons.scss'; -- can be an imported CSS stylesheet as well)
        use: [
          'style-loader',
          'css-loader', // no source map, no css module imports and no class renaming
          'sass-loader',
        ],
      },
    ],
  },
  ...
}

```

---

Example on how to use:

**src/styles/globals/globals.scss**

```
@import '~bootstrap/dist/css/bootstrap.min.css';
@import '~font-awesome/css/font-awesome.css';
```

**src/styles/variables/_variables.scss**

```
$navyblue: #114678;
```

**src/components/App/App.scss**

```
@import '../../styles/variables/variables';

App {
  color: $navyblue;
}
```

**src/components/App/App.js** (`globals.scss` can be imported in `src/index.js` instead)

```
import { App } from './App.scss';
import '../../styles/globals/globals.scss';

<div className={App}>
  <h1>Example</h1>
  <button className="btn btn-primary">
    Button
  </button>
  <i className="fa fa-fort-awesome" aria-hidden="true" />
</div>
```

**Comment by Laurencepublisher** [comment]
Check this https://nodejser.space/questions/what-is-the-reason-for-the-presence-of-the-symbol-in-node-module-names out for clarity
